packages/core/strapi/lib/services/server/middleware.js

```diff
@@ -1,28 +1,115 @@
 'use strict';
 
-const { propOr } = require('lodash/fp');
+const path = require('path');
+const { propOr, isArray } = require('lodash/fp');
 
 const getMiddlewareConfig = propOr([], 'config.middlewares');
 
-const resolveMiddlewares = route => {
+const resolveRouteMiddlewares = (route, strapi) => {
   const middlewaresConfig = getMiddlewareConfig(route);
 
-  return middlewaresConfig.map(middlewareConfig => {
-    if (typeof middlewareConfig === 'function') {
-      return middlewareConfig;
+  if (!isArray(middlewaresConfig)) {
+    throw new Error('Route middlewares config must be an array');
+  }
+
+  const middlewares = resolveMiddlewares(middlewaresConfig, strapi);
+
+  return middlewares.map(({ handler }) => handler);
+};
+
+/**
+ * Initialize every configured middlewares
+ * @param {MiddlewaresConfig} config
+ * @param {Strapi} strapi
+ * @returns {Middlewares}
+ */
+const resolveMiddlewares = (config, strapi) => {
+  const middlewares = [];
+
+  for (const item of config) {
+    if (typeof item === 'function') {
+      middlewares.push({
+        name: null,
+        handler: item,
+      });
+
+      continue;
+    }
+
+    if (typeof item === 'string') {
+      const middlewareFactory = strapi.middleware(item);
+
+      if (!middlewareFactory) {
+        throw new Error(`Middleware ${item} not found.`);
+      }
+
+      middlewares.push({
+        name: item,
+        handler: middlewareFactory({}, { strapi }),
+      });
+
+      continue;
     }
 
-    // TODO: this won't work until we have the new middleware formats
-    const middleware = strapi.middleware(middlewareConfig);
+    if (typeof item === 'object' && item !== null) {
+      const { name, resolve, config = {} } = item;
+
+      if (name) {
+        const middlewareFactory = strapi.middleware(name);
+        middlewares.push({
+          name,
+          handler: middlewareFactory(config, { strapi }),
+        });
+
+        continue;
+      }
```

Comment:
wouln't people want to put a name with the resolve ? That would break the logic. However by check resolve first, it would not. What do you think?